llvm/lib/Target/AIE/AIEBaseSubtarget.cpp

```diff
@@ -359,6 +479,64 @@ class EmitFixedSUnits : public ScheduleDAGMutation {
           AIE::maxLatency(&MI, *TII, *ItinData, /*IncludeStages=*/true));
       FixedDepSU->addPred(Dep, /*Required=*/true);
     }
+
+    // We only need to focus on top-fixed instructions when there is an Epilogue
+    // block.
+    if (BS.Kind != BlockType::Epilogue)
+      return;
+
+    MachineBasicBlock *Loop = AIELoopUtils::getLoopPredecessor(*DAG->getBB());
+    assert(Loop);
+    const BlockState &LBS = Scheduler->getInterBlock().getBlockState(Loop);
+    assert(LBS.Kind == BlockType::Loop);
+
+    if (!LBS.isPipelined())
+      return;
+
+    ArrayRef<AIE::MachineBundle> LoopTimedBundles = LBS.getTop().Bundles;
+    ArrayRef<AIE::MachineBundle> TopFixedBundles =
+        CurRegion.getTopFixedBundles();
+
+    RET.computeUseDefForward(TopFixedBundles, /*PastTheEndCycles*/ false);
+    // It is more cost-effective to reuse the RET to establish individual safety
+    // margins between the pipelined loop and the free instructions. This
+    // approach allows us to manage all dependencies related to EntrySU in one
+    // centralized location. While it is possible to implement this as a
+    // separate mutator, doing so could be costly, as it would prevent the
+    // creation of multiple edges from EntrySU to each free instruction that
+    // depends on both timed regions (TopFixed and LoopTimed).
+    RET.computeUseDefForward(LoopTimedBundles, /*PastTheEndCycles*/ true);
+
+    auto IsNonTopFixedSU = [Scheduler](const SUnit &SU) {
+      return !Scheduler->isFixedSU(SU, /*IsTop*/ true);
+    };
+
+    // Establish dependencies for each non top-fixed sched. unit by taking into
+    // account the def/use cycle of each operand.
+    for (SUnit &SU : make_filter_range(DAG->SUnits, IsNonTopFixedSU)) {
+      const MachineInstr &MI = *SU.getInstr();
+      if (const unsigned Latency = RET.getMaxOperandsDistance(MI)) {
+        SDep Dep(&DAG->EntrySU, SDep::Artificial);
+        Dep.setLatency(Latency);
+        SU.addPred(Dep, /*Required=*/true);
+      }
+    }
+
+    auto IsTopFixedSU = [Scheduler](const SUnit &SU) {
+      return Scheduler->isFixedSU(SU, true);
+    };
+
+    // TODO: this is pessimistic, we can handle this in RegionEndEdges after
```

Comment:
If we reorder, we need to take special care because it can break bot-insert instructions.